RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection
+{
+    /// <summary>
+    /// Interface for data collectors add-ins that choose to reprocess generated attachments
+    /// </summary>
+    public interface IDataCollectorAttachmentProcessor
+    {
+        /// <summary>
+        /// Gets the attachments Uris, which are handled by processor
+        /// </summary>
+        IEnumerable<Uri> GetExtensionUris();
+
+        /// <summary>
+        /// Indicates whether ProcessAttachmentSets is associative (e.g. f([a, b, c]) = f([f([a, b]), c]) = f([a, f([b, c])]))
+        /// </summary>
+        bool IsAssociative  { get; }
+
+        /// <summary>
+        /// Reprocess attachments generated by independent test executions
+        /// </summary>
+        /// <param name="attachments">Attachments to be processed</param>
+        /// <param name="progressReporter">Progress reporter. Accepts integers from 0 to 100</param>
+        /// <param name="logger">Message logger</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Attachments after reprocessing</returns>
+        ICollection<AttachmentSet> ProcessAttachmentSets(ICollection<AttachmentSet> attachments, IProgress<int> progressReporter, IMessageLogger logger, CancellationToken cancellationToken);
+    }
+}
+```
+
+Method `GetExtensionUris` should provide all Uris for data collector attachments which are handled by current processor. Test platform will provide to processor only data collector attachments with such Uris. Result of method `ProcessAttachmentSets` shouyld contain only data collector attachments with such Uris.
+
+`IsAssociative` should indicate if `ProcessAttachmentSets` is [associative](https://en.wikipedia.org/wiki/Associative_property).
+
+If `IsAssociative` is `True` Test Platform may try to speed up whole process by reprocessing data collector attachments as soon as possible when any two test executions are done. For example let's assume we have 5 test executions which are generating 5 data collector attachments: `a1`, `a2`, `a3`, `a4` and `a5`. Test platform could perform invocations:
+* `var result1 = ProcessAttachmentSets([a1, a2, a3], ...);` when first 3 executions are done
+* `var result2 = ProcessAttachmentSets(result1.Concat([a4]), ...);` when 4th execution is done
+* `var finalResult = ProcessAttachmentSets(result2.Concat([a5]), ...);` when last test execution is done
+
+If `IsAssociative` is `False` then Test Platform will wait for all test executions to finish and call `ProcessAttachmentSets` only once:
+* `var finalResult = ProcessAttachmentSets([a1, a2, a3, a4, a5], ...);`
+
+By default `IsAssociative` should be `False`, unless processing can take longer time and it's beneficial to start the process as soon as possible.
```

Comment:
multiTestRunCompleted  is exactly used for this purpose. TP will decide which attachment processors will be used based on this. multiTestRunCompleted  means that all test executions are done and we send to finalization all attachments.
@AbhitejJohn you were also proposing 'isLastBatch'. Do we want to rename it to 'isLastBatch'

I don't mind changing the naming here but I would like to agree finally on something as we have big PR in TP:
https://github.com/microsoft/vstest/pull/2463
and naming will hit it badly :)

Currently we use FinalizeMultiTestRun
Whole process is: Multi Test Run Finalization

@AbhitejJohn WDYT?

Reply:
The attachments may not be 'merged' from a processor standpoint. They could be split apart or additional metadata could be added or removed from them. A better term would be 'processed'. `ProcessTestRunAttachments` sounds like a good alternative to me given that the only functionality we are providing is exactly that - processing test run attachments across runs. I'm thinking the 'multi' part can be skipped given the scenario where processors could split/add/remove contents of a report. Given that, I'm not opposed to moving to `ProcessTestRunAttachments` if that makes sense to you as well.

Reply:
Functionally we should also make sure that the default code coverage processors do not throw if only one attachment is sent to them. They should ideally just return it back as is. I think you might already be doing so but just wanted to be explicit about it here.

Reply:
Regarding renaming we have now:

ProcessTestRunAttachments 

And whole process can be Test Run Attachments Processing or we can have Test Run Attachments Finalization

ITestRunAttachmentsProcessingEventsHandler  or ITestRunAttachmentsFinalizationEventsHandler

HandleTestRunAttachmentsProcessingComplete or HandleTestRunAttachmentsFinalizationComplete
HandleTestRunAttachmentsProcessingProgress or HandleTestRunAttachmentsFinalizationProgress

TestRunAttachmentsProcessingCompleteEventArgs or TestRunAttachmentsFinalizationCompleteEventArgs 
TestRunAttachmentsProcessingProgressEventArgs or TestRunAttachmentsFinalizationProgressEventArgs 

@AbhitejJohn @drognanar please reply with final proposal so I can do changes in TP :) I will consider writing simple script to go and rename everything :D

Reply:
After more thinking about this maybe we could leave current naming: FinalizeMultiTestRun, MultiTestRunFinalization.
1) maybe in future we will want to merge/combine something else than attachments. Internally we can still use all functions and classes with current names.
2) finalize indicates that this is last thing of some process (test runs in this case). For someone who is reading our interface it could be easier to understand that this should be used after test runs

Reply:
Renaming done